### PR TITLE
f-AdvancedIIATesting

### PR DESCRIPTION
**Pull request by pdschubert**

### CODE REVIEW

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h [hunk 1]
```diff
@@ -670,9 +734,56 @@ class IDEInstInteractionAnalysisT
...
+    l_t UserEdgeFacts = BitVectorSet<e_t>();
+    std::set<e_t> EdgeFacts;
+    if (edgeFactGen) {
+      EdgeFacts = edgeFactGen(callSite);
+      // fill BitVectorSet
+      UserEdgeFacts = BitVectorSet<e_t>(EdgeFacts.begin(), EdgeFacts.end());
+    }
+    // Model call to heap allocating functions (new, new[], malloc, etc.) --
+    // only model direct calls, though.
+    if (callees.size() == 1) {
```
Comment (vulder) [inline include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h:747]: What happens in the case we have non direct calls, maybe through an allocation? Not sure if this is correctly handled by the other code.

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h [hunk 2]
```diff
@@ -253,26 +261,28 @@ class IDEInstInteractionAnalysisT
...
-              Facts.erase(src);
-            }
+            // If the value to be stored does not hold we must at least add the
+            // store instruction and the points-to set as the instruction still
+            // interacts with the memory locations pointed to be PTS.
+            // if (Store->getPointerOperand() == src ||
+            //        PointerPTS->count(src)) {
+            //   Facts.insert(Store);
+            //  Facts.erase(src);
+            // }
```
Comment (vulder) [inline include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h:275]: Why is this commented? Not needed or did you miss replacing it?

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h [hunk 3]
```diff
@@ -147,30 +150,32 @@ class IDEInstInteractionAnalysisT
...
+    //           // this->indirecrTaints[succ].insert(src);
+    //           for (auto &inst : succ->instructionsWithoutDebug()) {
+    //             ret.insert(&inst);
+    //           }
+    //         }
+    //       }
+    //       return ret;
+    //     });
+    //   }
+    // }
```
Comment (vulder) [inline include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h:171]: Hmm, I think this should be readded or am I wrong? Or do we handle the propagation of indirect taints somewhere else?

**Review by vulder — CHANGES_REQUESTED** [review]
Found just a few things we should clean up before merging.
:+1: for the comments, makes it much easier to follow what the analysis should be doing.

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h [hunk 4]
```diff
@@ -936,28 +1047,38 @@ class IDEInstInteractionAnalysisT
...
-    if (Lhs == BottomElement || Rhs == BottomElement) {
-      return BottomElement;
-    }
     if (Lhs == TopElement) {
       return Rhs;
     }
     if (Rhs == TopElement) {
       return Lhs;
     }
+    if (Lhs == BottomElement) {
```
[the remark this answers is not in the document]
Comment (pdschubert) [reply include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h:1062]: I don't have hard feelings about that. Will resort to the `or` variant.

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h (hunk 1 above)
Comment (pdschubert) [reply include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h:743]: Let's discuss that in our next meeting.

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h (hunk 3 above)
Comment (pdschubert) [reply include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h:167]: Added it again. There is also a bug in the code which I fixed on the fly.

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h (hunk 2 above)
Comment (pdschubert) [reply include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h:271]: I need to double check on that one. There was some issue here ...

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h [hunk 5]
```diff
@@ -146,25 +146,48 @@ class IDEInstInteractionAnalysisT
...
-    //           }
-    //         }
-    //       }
-    //       return ret;
-    //     });
-    //   }
-    // }
+    // Handle indirect taints, i. e., propagate values that depend on branch
+    // conditions whose operands are tainted.
+    if (EnableIndirectTaints) {
```
Comment (vulder) [inline include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h:151]:
```suggestion
    if constexpr (EnableIndirectTaints) {
```
Should not matter much but putting constexpr here makes it clear that EnableIndirectTaints is a compile time constant and the block will not be compiled in when it is `false`

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h [hunk 6]
```diff
@@ -146,25 +146,48 @@ class IDEInstInteractionAnalysisT
...
+      if (auto Br = llvm::dyn_cast<llvm::BranchInst>(curr);
```
Comment (vulder) [inline include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDEInstInteractionAnalysis.h:152]:
```suggestion
     if (const auto *Br = llvm::dyn_cast<llvm::BranchInst>(curr);
```

**Review by vulder — CHANGES_REQUESTED** [review]

**Comment by vulder** [comment]
@pdschubert 
One thing I missed, I think we should extend the tests for the analysis to capture the problems we found so we can prevent them from being reintroduced.

**Review by vulder — APPROVED** [review]
Fixes work for my example codes :+1:

**Pull request merged**